File: apps/clients/android-app/app/src/main/java/com/streamverse/mobile/viewmodel/ContentViewModel.kt
package com.streamverse.mobile.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.streamverse.mobile.data.repository.ContentRepository
import com.streamverse.mobile.models.Content
import com.streamverse.mobile.models.ContentRow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for content operations.
 */
class ContentViewModel(
    private val contentRepository: ContentRepository
) : ViewModel() {

    private val _contentRows = MutableStateFlow<List<ContentRow>>(emptyList())
    val contentRows: StateFlow<List<ContentRow>> = _contentRows.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    /**
     * Load home content.
     */
    fun loadHomeContent() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                _contentRows.value = contentRepository.getHomeContentRows()
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to load content"
                _contentRows.value = emptyList()
            } finally {
                _isLoading.value = false
            }
        }
    }

    /**
     * Search content.
     */
    fun search(query: String, onResult: (List<Content>) -> Unit) {
        if (query.length < 2) {
            onResult(emptyList())
            return
        }

        viewModelScope.launch {
            try {
                val results = contentRepository.searchContent(query)
                onResult(results)
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Search failed"
                onResult(emptyList())
            }
        }
    }

    /**
     * Get content by ID.
     */
    fun getContentById(id: String, onResult: (Content?) -> Unit) {
        viewModelScope.launch {
            try {
                val content = contentRepository.getContentById(id)
                onResult(content)
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to load content"
                onResult(null)
            }
        }
    }

    class Factory(
        private val repository: ContentRepository
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ContentViewModel(repository) as T
        }
    }
}

File: apps/clients/ios-app/StreamVerse/ViewModels/ContentViewModel.swift
//
//  ContentViewModel.swift
//  StreamVerse
//
//  Created on 2025-01-XX.
//

import Foundation
import Combine

/// ViewModel for content operations.
@MainActor
class ContentViewModel: ObservableObject {
    private let contentService = ContentService.shared
    
    @Published var contentRows: [ContentRow] = []
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    /// Load home content.
    func loadHomeContent() async {
        isLoading = true
        errorMessage = nil
        
        do {
            contentRows = try await contentService.getHomeContent()
        } catch {
            errorMessage = error.localizedDescription
            contentRows = []
        }
        
        isLoading = false
    }
    
    /// Search content.
    func search(query: String) async -> [Content] {
        guard query.count >= 2 else { return [] }
        
        do {
            return try await contentService.searchContent(query: query)
        } catch {
            errorMessage = error.localizedDescription
            return []
        }
    }
    
    /// Get content by ID.
    func getContent(byId id: String) async -> Content? {
        do {
            return try await contentService.getContentById(id)
        } catch {
            errorMessage = error.localizedDescription
            return nil
        }
    }
}

